src/neo/Ledger/Blockchain.cs

```diff
@@ -263,22 +266,57 @@ private void OnImport(IEnumerable<Block> blocks, bool verify)
             Sender.Tell(new ImportCompleted());
         }
 
+        private void RemoveUnverifiedBlockToCache(uint index, bool removeKnownHashes)
+        {
+            if (block_cache_unverified.Remove(index, out var entry))
+            {
+                block_cache_unverified_size -= entry.Size;
+
+                if (removeKnownHashes)
+                {
+                    var hashes = entry.Blocks.Select(u => u.Hash).ToArray();
+
+                    system.LocalNode.Tell(new TaskManager.RemoveKnownHashes { Hashes = hashes });
+                    system.TaskManager.Tell(new TaskManager.RemoveKnownHashes { Hashes = hashes });
+                }
+            }
+        }
+
         private void AddUnverifiedBlockToCache(Block block)
         {
+            if (block_cache_unverified_size > MaxUnverifiedBlockSize && block.Index > Height + 10)
```

Comment:
This logic does not take malfactor into account. Suppose there is a malfactor neighbor who keeps on sending fake blocks whose index is lower than Height + 10. By current logic all these blocks will be stored and func RemoveUnverifiedBlockToCache might never be executed.

Reply:
Fixed

Reply:
I think we should firstly add and than execute RemoveUnverifiedBlockToCache to avoid the situation that deleted index is lower than the index of new coming block?

Reply:
Then we can exceed the memory limit

Reply:
And still malicious neighbor can flush out normal unverified blocks from other neighbor i.e. by bombarding lots of fake blocks with index = Height + 1. Maybe we can add a filter in RemoteNode to avoid such occasion?

Reply:
Or we can leave this to #1811

Reply:
Ban nodes should be done in #1811